src/Dom/DocumentMiddleware.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace AmpProject\Dom;
+
+use AmpProject\Dom\Document;
+use DOMNode;
+
+interface DocumentMiddleware
+{
+
+    /**
+     * Preprocess the HTML to be loaded into the Dom\Document.
+     *
+     * @param string   $html     String of HTML markup to be preprocessed.
+     * @param Document $document Document to be preprocessed.
+     * @return string Preprocessed string of HTML markup.
+     */
+    public function beforeLoad($html, Document $document);
+
+    /**
+     * Preprocess the DOM to be saved into HTML.
+     *
+     * @param string   $html     String of HTML markup to be preprocessed.
+     * @param Document $document Document to be preprocessed before saving it into HTML.
+     * @param DOMNode  $node     Optional. Parameter to output a subset of the document.
+     * @return string Preprocessed string of HTML markup.
+     */
+    public function beforeSave($html, Document $document, DOMNode $node = null);
```

Comment:
The interface is fuzzy, because these methods handle two different points in time each. That's why you're dealing with both THML strings and documents in both.

As I already briefly mentioned during our last meeting, I think the complete interface will need to have 4 methods:
1. `beforeLoad(string $html): string` dealing with HTML before it is parsed into a DOM;
2. `afterLoad(Document $document): void` dealing with the DOM after it has been parsed from the HTML;
3. `beforeSave(Document $document):void` dealing with the DOM before it is being rendered into HTML again;
4. `afterSave(string $html): string` dealing with the HTML after it has been rendered out of the final DOM.
(Note that most of the types are only here for communication the intent, but should be stripped as needed in the code to stick to PHP 5.6).

Given that most middleware will only use 1-2 of these methods, I think we should turn this from an interface into an abstract base class, where the 4 methods are already implemented but do nothing. Each middleware will then implement only those that it needs.

Then, when you build an implementation, you can actually put the logic into the method you're overriding. So, instead of this:
```php
public function beforeLoad($html) {
	return $this->secureSomething($html);
}
private function secureSomething($html) {
	// logic to manipulate $html.
	return $html;
}
```
we can use this instead:
```php
public function beforeLoad($html) {
	// logic to manipulate $html.
	return $html;
}
```